### PR TITLE
Case 21494: Avoid mem leaked entities

### PR DESCRIPTION
**Pull request by AndrewMeadows**
https://highfidelity.manuscript.com/f/cases/21494/Entity-Collisions-from-previously-visited-domains-follow-you-to-new-domains-0-81

This PR fixes a mem-leak and also ghost collision shapes from previous regions in the physics simulation.

### CODE REVIEW

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -3773,9 +3773,12 @@ std::map<QString, QString> Application::prepareServerlessDomainContents(QUrl dom
         tmpTree->reaverageOctreeElements();
         tmpTree->sendEntities(&_entityEditSender, getEntities()->getTree(), 0, 0, 0);
     }
+    std::map<QString, QString> namedPaths = tmpTree->getNamedPaths();
 
-    return tmpTree->getNamedPaths();
+    // we must manually eraseAllOctreeElements(false) else the tmpTree will mem-leak
+    tmpTree->eraseAllOctreeElements(false);
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: This fixes a mem-leak from loading serverless domains.

File: libraries/entities/src/EntityTree.cpp [hunk 2]
```diff
@@ -52,7 +52,15 @@ EntityTree::EntityTree(bool shouldReaverage) :
...
-    eraseAllOctreeElements(false);
+    // NOTE: to eraseAllOctreeElements() in this context is useless because
+    // any OctreeElements in the tree still have shared backpointers to this Tree
+    // which means the dtor never would have been called in the first place!
+    //
+    // I'm keeping this useless commented-out line to remind us:
+    // we don't need shared pointer overhead for EntityTrees.
+    // TODO: EntityTreeElement::_tree should be raw back pointer.
+    // AND: EntityItem::_element should be a raw back pointer.
+    //eraseAllOctreeElements(false); // KEEP THIS
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityTree.cpp]: Surprise!  `EntityTree` requires `std::shared_ptr` usage, and there are hidden references to the `EntityTree` inside the `EntityTreeElement`s, which have hidden references in their EntityItems.  This all means: the `EntityTree::dtor` will not be called unless/until external code manually calls `EntityTree::eraseAllOctreeElements(false)`.

File: libraries/entities/src/EntityTree.cpp [hunk 3]
```diff
@@ -128,6 +140,11 @@ void EntityTree::eraseAllOctreeElements(bool createNewRoot) {
             if (element) {
                 element->cleanupEntities();
             }
+            int32_t spaceIndex = entity->getSpaceIndex();
+            if (spaceIndex != -1) {
+                // assume stale spaceIndices will be freed later
+                _staleProxies.push_back(spaceIndex);
+            }
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityTree.cpp]: This fixes the other mem-leak: don't `_staleProxies.clear()` higher up and instead keep the list of indices around so they can be freed later (which will subsequentlty release some `EntityItemPointer` references).

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17749/HighFidelity-Beta-PR15060-8a7853a.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17749/HighFidelity-Quest-Beta-PR15060-8a7853a.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17749/HighFidelity-Beta-PR15060-8a7853a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17749/HighFidelity-Beta-Interface-PR15060-8a7853a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17749/HighFidelity-Beta-PR15060-8a7853a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17749/HighFidelity-Beta-Interface-PR15060-8a7853a.exe

**Pull request merged**